File: src/parse.js
import { adapters } from './adapters/index.js';
import { extractScheme } from './adapters/common.js';
import { isS3HttpUrl } from './adapters/s3.js';
import { diagnostic, fail } from './diagnostics.js';
import { looksLikeFilePath } from './path.js';
import { createRegistry, defaultRegistry, getBuiltInDefinitions } from './registry.js';
import { mask } from './redaction.js';
import { validateAddress } from './validation.js';

function ok(value, errors = [], warnings = []) {
  return {
    ok: errors.length === 0,
    value: errors.length === 0 ? value : null,
    errors,
    warnings
  };
}

function normalizeInput(input) {
  if (typeof input !== 'string') {
    return { error: fail('INVALID_INPUT_TYPE', 'Connparse input must be a string') };
  }
  if (!input.trim()) {
    return { error: fail('EMPTY_INPUT', 'Connparse input cannot be empty') };
  }
  return { raw: input };
}

function inferDefinition(raw, registry) {
  if (isS3HttpUrl(raw)) {
    return { scheme: 's3', definition: registry.getByScheme('s3') };
  }

  const scheme = extractScheme(raw);
  if (!scheme && looksLikeFilePath(raw)) {
    return { scheme: 'file', definition: registry.getByScheme('file') };
  }

  if (!scheme) {
    return { scheme: null, definition: null };
  }

  return { scheme, definition: registry.getByScheme(scheme) };
}

function parseUnknown(raw, scheme, strict) {
  const warning = diagnostic('UNKNOWN_SCHEME', `${scheme} does not have a registered Connparse definition`, 'scheme');
  if (strict) return fail(warning.code, warning.message, warning.path);

  try {
    const url = new URL(raw);
    return ok(
      {
        scheme,
        type: 'unknown',
        authority: {
          host: url.hostname,
          port: url.port ? Number(url.port) : null
        },
        resource: {
          type: 'unknown',
          name: url.pathname ? decodeURIComponent(url.pathname.replace(/^\//, '').split('/')[0] || '') : null
        },
        path: decodeURIComponent(url.pathname || ''),
        query: Object.fromEntries(new URLSearchParams(url.search)),
        fragment: url.hash ? decodeURIComponent(url.hash.slice(1)) : null,
        credentials: {},
        options: {},
        raw,
        safe: mask(raw)
      },
      [],
      [warning]
    );
  } catch {
    return fail('INVALID_URL', `Could not parse ${scheme} address`, 'raw');
  }
}

export function parse(input, options = {}) {
  const normalized = normalizeInput(input);
  if (normalized.error) return normalized.error;

  const raw = normalized.raw;
  const registry = options.definitions
    ? createRegistry([...getBuiltInDefinitions(), ...options.definitions])
    : defaultRegistry;
  const { scheme, definition } = inferDefinition(raw, registry);

  if (!scheme) {
    return fail('MISSING_SCHEME', 'Input must include a scheme or look like a file path', 'scheme');
  }

  if (!definition) {
    return parseUnknown(raw, scheme, Boolean(options.strict));
  }

  const adapterName = definition.adapter || 'generic-uri';
  const adapter = adapters[adapterName];
  if (!adapter) {
    return fail('UNKNOWN_ADAPTER', `${definition.id} references missing adapter ${adapterName}`, 'adapter');
  }

  try {
    const value = adapter(raw, definition, {
      raw,
      safe: mask(raw)
    });
    const validation = validateAddress(value, definition, options);
    return ok(value, validation.errors, validation.warnings);
  } catch (error) {
    return fail('PARSE_FAILED', error instanceof Error ? error.message : String(error), 'raw');
  }
}

export function parseOrThrow(input, options = {}) {
  const result = parse(input, options);
  if (result.ok) return result.value;
  const message = result.errors.map((error) => error.message).join('; ');
  throw new Error(message || 'Connparse failed');
}

File: packages/go/parse.go
package connparse

import (
	"errors"
	"fmt"
	"regexp"
	"strings"
)

type adapterFunc func(string, Definition, string) (*Address, error)

var adapters = map[string]adapterFunc{
	"generic-uri":         parseGenericURI,
	"mongodb":             parseMongoDB,
	"object-storage":      parseObjectStorage,
	"redis":               parseRedis,
	"file":                parseFile,
	"sqlite":              parseSQLite,
	"duckdb":              parseDuckDB,
	"jdbc":                parseJDBC,
	"clickhouse":          parseClickHouse,
	"memcached":           parseMemcached,
	"elasticsearch":       parseElasticsearch,
	"s3":                  parseS3,
	"questdb":             parseQuestDB,
	"postgres-compatible": parsePostgresCompatible,
	"mysql-compatible":    parseMySQLCompatible,
}

func Parse(input string, options ...Options) Result {
	opts := Options{}
	if len(options) > 0 {
		opts = options[0]
	}
	if strings.TrimSpace(input) == "" {
		return fail("EMPTY_INPUT", "Connparse input cannot be empty", "raw")
	}

	definitions := BuiltInDefinitions()
	definitions = append(definitions, opts.Definitions...)
	registry, err := NewRegistry(definitions)
	if err != nil {
		return fail("INVALID_DEFINITION", err.Error(), "definitions")
	}

	scheme, def, found := inferDefinition(input, registry, opts)
	if scheme == "" {
		return fail("MISSING_SCHEME", "Input must include a scheme or look like a file path", "scheme")
	}
	if !found {
		return parseUnknown(input, scheme, opts.Strict)
	}

	adapterName := def.Adapter
	if adapterName == "" {
		adapterName = "generic-uri"
	}
	adapter, adapterFound := adapters[adapterName]
	if !adapterFound {
		return fail("UNKNOWN_ADAPTER", fmt.Sprintf("%s references missing adapter %s", def.ID, adapterName), "adapter")
	}

	value, err := adapter(input, def, input)
	if err != nil {
		return fail("PARSE_FAILED", err.Error(), "raw")
	}

	validation := validateAddress(value, def, opts)
	return ok(value, validation.errors, validation.warnings)
}

func ParseOrThrow(input string, options ...Options) (*Address, error) {
	result := Parse(input, options...)
	if result.OK {
		return result.Value, nil
	}
	if len(result.Errors) == 0 {
		return nil, errors.New("Connparse failed")
	}
	messages := make([]string, len(result.Errors))
	for i, item := range result.Errors {
		messages[i] = item.Message
	}
	return nil, errors.New(strings.Join(messages, "; "))
}

func inferDefinition(raw string, registry *Registry, options Options) (string, Definition, bool) {
	if options.Provider != "" {
		provider := strings.ToLower(options.Provider)
		if def, ok := registry.ByID(provider); ok {
			return provider, def, true
		}
		if def, ok := registry.ByScheme(provider); ok {
			return provider, def, true
		}
		return provider, Definition{}, false
	}

	if isClickHouseJDBC(raw) {
		def, ok := registry.ByScheme("clickhouse")
		return "jdbc:clickhouse", def, ok
	}
	if regexp.MustCompile(`(?i)^jdbc:postgresql:\/\/`).MatchString(raw) {
		def, ok := registry.ByScheme("postgres")
		return "jdbc:postgresql", def, ok
	}
	if regexp.MustCompile(`(?i)^jdbc:mysql:\/\/`).MatchString(raw) {
		def, ok := registry.ByScheme("mysql")
		return "jdbc:mysql", def, ok
	}
	if regexp.MustCompile(`(?i)^jdbc:mariadb(?::[a-z-]+)?:\/\/`).MatchString(raw) {
		def, ok := registry.ByScheme("mariadb")
		return "jdbc:mariadb", def, ok
	}
	if isQuestDBConfig(raw) {
		def, ok := registry.ByScheme("questdb")
		return "questdb", def, ok
	}
	if looksLikeDuckDBPath(raw) {
		def, ok := registry.ByScheme("duckdb")
		return "duckdb", def, ok
	}
	if isS3HTTPURL(raw) {
		def, ok := registry.ByScheme("s3")
		return "s3", def, ok
	}

	scheme := extractScheme(raw)
	if scheme == "" && looksLikeFilePath(raw) {
		def, ok := registry.ByScheme("file")
		return "file", def, ok
	}
	if scheme == "" {
		return "", Definition{}, false
	}
	def, ok := registry.ByScheme(scheme)
	return scheme, def, ok
}

func parseUnknown(raw string, scheme string, strict bool) Result {
	warning := diagnostic("UNKNOWN_SCHEME", fmt.Sprintf("%s does not have a registered Connparse definition", scheme), "scheme")
	if strict {
		return fail(warning.Code, warning.Message, warning.Path)
	}

	p, err := parseHierarchical(raw)
	if err != nil {
		return fail("INVALID_URL", fmt.Sprintf("Could not parse %s address", scheme), "raw")
	}
	name, _ := firstRest(p.PathSegments)
	value := baseAddress(
		Definition{Type: "unknown"},
		scheme,
		raw,
		Mask(raw),
		map[string]any{"host": p.Host, "port": p.Port},
		Resource{Type: "unknown", Name: nullable(name)},
		p.Pathname,
		p.Query,
		p.Fragment,
		nil,
		nil,
	)
	return ok(value, nil, []Diagnostic{warning})
}
